pkg/invariants/incluster_disruption_serializer/remote.go

```diff
@@ -381,18 +285,181 @@ func fetchEventsFromFileOnNode(ctx context.Context, clientset *kubernetes.Client
+}
+
+func (i *InvariantInClusterDisruption) StartCollection(ctx context.Context, adminRESTConfig *rest.Config, _ monitorapi.RecorderWriter) error {
+	fmt.Fprintf(os.Stderr, "Starting in-cluster monitoring daemonsets\n")
+	i.adminRESTConfig = adminRESTConfig
+	kubeClient, err := kubernetes.NewForConfig(i.adminRESTConfig)
+	if err != nil {
+		return err
+	}
+	configClient, err := configclient.NewForConfig(i.adminRESTConfig)
+	if err != nil {
+		return err
+	}
+	infra, err := configClient.ConfigV1().Infrastructures().Get(context.Background(), "cluster", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	internalAPI, err := url.Parse(infra.Status.APIServerInternalURL)
+	if err != nil {
+		return err
+	}
+	apiIntHost := internalAPI.Hostname()
+
+	err = createNamespace(ctx, kubeClient)
+	if err != nil {
+		return err
+	}
+	err = createServiceAccount(ctx, kubeClient)
+	if err != nil {
+		return err
+	}
+	err = createRBACPrivileged(ctx, kubeClient)
+	if err != nil {
+		return err
+	}
+	err = createMonitorRole(ctx, kubeClient)
+	if err != nil {
+		return err
+	}
+	err = createMonitorCRB(ctx, kubeClient)
+	if err != nil {
+		return err
+	}
+	err = createServiceNetworkDS(ctx, kubeClient)
+	if err != nil {
+		return err
+	}
+	err = createLocalhostDS(ctx, kubeClient)
+	if err != nil {
+		return err
+	}
+	err = createInternalLBDS(ctx, kubeClient, apiIntHost)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (i *InvariantInClusterDisruption) CollectData(ctx context.Context, storageDir string, beginning time.Time, end time.Time) (monitorapi.Intervals, []*junitapi.JUnitTestCase, error) {
+	fmt.Fprintf(os.Stderr, "Stopping in-cluster monitoring daemonsets\n")
+	client, err := kubernetes.NewForConfig(i.adminRESTConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Start watcher before daemonset removal request to make sure delete event is not skipped
+	c := make(chan error)
+	timeLimitedCtx, cancel := context.WithTimeout(ctx, 5*time.Minute)
+	defer cancel()
+	go func() {
+		_, watchErr := watchtools.UntilWithSync(timeLimitedCtx,
+			cache.NewListWatchFromClient(
+				client.AppsV1().RESTClient(), "daemonsets", namespace, fields.Everything()),
+			&appsv1.DaemonSet{},
+			nil,
+			func(event watch.Event) (bool, error) {
+				return event.Type == watch.Deleted, nil
+			},
+		)
+		if watchErr != nil {
+			watchErr = fmt.Errorf("daemonsets in namespace %s didn't get destroyed: %v", namespace, watchErr)
+		}
+		c <- watchErr
+	}()
+
+	// Remove daemonsets first to avoid trailing false-positive disruption intervals
+	dsClient := client.AppsV1().DaemonSets(namespace)
+	err = dsClient.DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, nil, fmt.Errorf("error removing daemonsets in namespace %s: %v", namespace, err)
+	}
+	if watchErr := <-c; watchErr != nil {
+		return nil, nil, watchErr
+	}
+
+	fmt.Fprintf(os.Stderr, "Fetching collected in-cluster disruption data\n")
```

Comment:
If you have a soft "stop collecting" (say when a configmap is created), then you can use pod logs instead of node extraction.  That avoids limiting you particular node directories not accessible on general kube deployments.

Reply:
We can't rely on pod logs, as previous pod logs will be gone during upgrade (or anything which removes a pod)

Reply:
I do like the idea of "stop when configmap created" (maybe some annotation on the pod would be easier?) - it makes debugging this much easier

Reply:
pod logs is a better solution as it won't require us to wait for DS to be fully stopped - and easier to port for upstream. See example in https://github.com/openshift/origin/pull/28176